src/main/io/gps.c

```diff
@@ -1337,14 +1341,53 @@ static void calculateNavInterval (void)
     gpsSol.navIntervalMs = constrain(navDeltaTimeMs, 50, 2500);
 }
 
+#if defined(USE_VIRTUAL_GPS)
+static void updateVirtualGPS(void)
+{
+    const uint32_t updateInterval = 100; // 100ms 10Hz update time interval
+    static uint32_t nextUpdateTime = 0;
+
+    if (cmp32(gpsData.now, nextUpdateTime) > 0) {
+        if (gpsData.state == GPS_STATE_INITIALIZED) {
+            gpsSetState(GPS_STATE_RECEIVING_DATA);
+        }
+
+        getVirtualGPS(&gpsSol);
+        gpsSol.time = gpsData.now;
+
+        gpsData.lastNavMessage = gpsData.now;
+        sensorsSet(SENSOR_GPS);
+
+        if (gpsSol.numSat > 3) {
+            gpsSetFixState(GPS_FIX);
+        } else {
+            gpsSetFixState(false);
+        }
+        GPS_update ^= GPS_DIRECT_TICK;
+
+        calculateNavInterval();
+        onGpsNewData();
+
+        nextUpdateTime = gpsData.now + updateInterval;
+    }
+}
+#endif
+
 void gpsUpdate(timeUs_t currentTimeUs)
 {
     static timeDelta_t gpsStateDurationFractionUs[GPS_STATE_COUNT];
     timeDelta_t executeTimeUs;
     gpsState_e gpsCurrentState = gpsData.state;
     gpsData.now = millis();
 
+#if defined(USE_VIRTUAL_GPS)
+    if (gpsConfig()->provider == GPS_VIRTUAL) {
+        updateVirtualGPS();
+    }
+    else
+#endif
```

Comment:
IMO it is cleaner to move it to line https://github.com/betaflight/betaflight/pull/14131/files#diff-15b828e75456a052109384cb9af7da6fb6001ebdf63c417647471ca2201cdee8R1437
That way it will allow easier extension.

```suggestion
#if defined(USE_VIRTUAL_GPS)
    } else if (gpsConfig()->provider == GPS_VIRTUAL) {
        updateVirtualGPS();
    }
#endif
```

Or ublox/nmea/msp/virtual may be changes into
```
switch (gpsConfig()->provider) {
case GPS_UBLOX:
case GPS_NMEA:
    if (!gpsPort) {
        break;
    }  
    DEBUG_SET(DEBUG_GPS_CONNECTION, 7, serialRxBytesWaiting(gpsPort));
    ....
    break;
case GPS_MSP:
    if (GPS_update & GPS_MSP_UPDATE) { 
   ... 
    break;
case GPS_VIRTUAL:
   ...